File: cc/libcc/nx/stats/stats.cc
//
// weproxy@foxmail.com 2022/10/03
//

#include "stats.h"

#include "fx/fx.h"
#include "gx/fmt/fmt.h"
#include "logx/logx.h"

namespace nx {
namespace stats {
namespace xx {

////////////////////////////////////////////////////////////////////////////////
//

connT _Conn;
rateT _Rate;
dnsT _DNS;

////////////////////////////////////////////////////////////////////////////////
//
string connE::String() const {
    return fmt::Sprintf("T:%lld/%lld,U:%lld/%lld", tcp.load(), tcpTotal.load(), udp.load(), udpTotal.load());
}

////////////////////////////////////////////////////////////////////////////////
//
void rateE::calc(Type typ, bool tcp) {
    int64 sentRate = sent.exchange(0);
    int64 recvRate = recv.exchange(0);

    if (sentRate > 0 || recvRate > 0) {
        LOGS_P("[rate] " << ToString(typ) << (tcp ? " TCP" : " UDP") << ": sent=" << fx::FormatBytes(sentRate)
                         << "/s, recv=" << fx::FormatBytes(recvRate) << "/s");
    }
}
////////////////////////////////////////////////////////////////////////////////
//
void dnsT::calc() {
    if (!dirty.exchange(false)) {
        return;
    }

    LOGS_P(TAG << " DNS{dist/indist}: total=" << distinct.total << "/" << all.total << ", succ=" << distinct.succeeded
               << "/" << all.succeeded << ", fail=" << distinct.failed << "/" << all.failed
               << ", hit=" << distinct.cahcehit << "/" << all.cahcehit << ", drop=" << distinct.dropped << "/"
               << all.dropped << ", fake=" << distinct.faked << "/" << all.faked);
}

////////////////////////////////////////////////////////////////////////////////
//
void connT::calc() {
    if (!dirty.exchange(false)) {
        return;
    }

    std::ostringstream oss;
    oss << TAG << " Conn{cur/his}: ";

    for (int i = 0; i < xx::typeCount; i++) {
        Type typ = (Type)(TypeDirect + i);
        if (typ == typeMax || (arr[i].tcpTotal == 0 || arr[i].tcpTotal == 0)) {
            continue;
        }

        oss << ToString(typ) << "{" << arr[i].String() << "} ";
    }

    LOGS_P(oss.str());
}
}  // namespace xx

// LogD ...
const Stats& Stats::LogD(const string& msg) const {
    LOGS_D(tag << " " << msg << ", " << Elapsed());
    return *this;
}

// LogI ...
const Stats& Stats::LogI(const string& msg) const {
    LOGS_I(tag << " " << msg << ", " << Elapsed());
    return *this;
}

// LogW ...
const Stats& Stats::LogW(const string& msg) const {
    LOGS_W(tag << " " << msg << ", " << Elapsed());
    return *this;
}

// LogE ...
const Stats& Stats::LogE(const string& msg) const {
    LOGS_E(tag << " " << msg << ", " << Elapsed());
    return *this;
}

// Start ...
Stats& Stats::Start(const string& msg) {
    start = time::Now();

    if (!msg.empty()) {
        LOGS_I(tag << " " << msg);
    }

    if (tcp) {
        conn.AddTCP(typ, 1);
    } else {
        conn.AddUDP(typ, 1);
    }

    return *this;
}

// Done ...
Stats& Stats::Done(const string& msg) {
    if (tcp) {
        conn.AddTCP(typ, -1);
    } else {
        conn.AddUDP(typ, -1);
    }

    if (sentx > 0 || recvx > 0) {
        LOGS_I(tag << " " << msg << ", sent=" << fx::FormatBytes(sentx) << "+" << fx::FormatBytes(sent) + ", recv="
                   << fx::FormatBytes(recvx) << "+" << fx::FormatBytes(recv) << ", " << Elapsed());
    } else {
        LOGS_I(tag << " " << msg << ", sent=" << fx::FormatBytes(sent) + ", recv=" << fx::FormatBytes(recv) << ", "
                   << Elapsed());
    }

    return *this;
}

////////////////////////////////////////////////////////////////////////////////
// init ...
namespace xx {
static auto _ = [] {
    gx::go([] {
        LOGS_V(TAG << " init()");
        for (;;) {
            _Rate.calc();
            _DNS.calc();
            _Conn.calc();
            time::Sleep(time::Second);
        }
    });
    return 0;
}();
}  // namespace xx
}  // namespace stats
}  // namespace nx

File: go/libgo/nx/stats/stats.go
//
// weproxy@foxmail.com 2022/10/20
//

package stats

import (
	"fmt"
	"sync/atomic"
	"time"
	"weproxy/acc/libgo/fx"
	"weproxy/acc/libgo/logx"
)

// Type ...
type Type int

const (
	TypeDirect Type = iota
	TypeS5
	TypeSS
	TypeGAAP
	TypeHTP
	TypeDNS
	TypeKCP
	TypeQUIC
	TypeOTHER
	typeMax
)

// typeCount ...
const typeCount = typeMax - TypeDirect

// String ...
func (m Type) String() string {
	return []string{"Direct", "S5", "SS", "GAAP", "HTP", "DNS", "KC", "QC", "OTH", ""}[m-TypeDirect]
}

// /////////////////////////////////////////////////////////////////////////////////////////////////
// Conn  ...
var Conn *connT = newConnT()

// Rate  ...
var Rate *rateT = newRateT()

// DNS  ...
var DNS *dnsT = newDnsT()

///////////////////////////////////////////////////////////////////////////////////////////////////

const TAG = "[stats]"

// init ...
func init() {
	go func() {
		t := time.NewTicker(time.Second)
		for range t.C {
			Rate.calc()
			DNS.calc()
			Conn.calc()
		}
	}()
}

///////////////////////////////////////////////////////////////////////////////////////////////////

// Stats ...
type Stats struct {
	sent, recv   int64
	sentx, recvx int64
	tcp          bool
	start        time.Time
	tag          string
	typ          Type
	conn         *connT
	rate         *rateT
	target       string
	server       string
}

// newStats ...
func newStats(typ Type, tag string, tcp bool) *Stats {
	return &Stats{
		tcp:  tcp,
		tag:  tag,
		typ:  typ,
		conn: Conn,
		rate: Rate,
	}
}

// NewTCPStats ...
func NewTCPStats(typ Type, tag string) *Stats {
	return newStats(typ, tag, true)
}

// NewUDPStats ...
func NewUDPStats(typ Type, tag string) *Stats {
	return newStats(typ, tag, false)
}

// Tag ...
func (m *Stats) Tag() string {
	return m.tag
}

// SetTag ...
func (m *Stats) SetTag(tag string) *Stats {
	m.tag = tag
	return m
}

// Target ...
func (m *Stats) Target() string {
	return m.target
}

// SetTarget ...
func (m *Stats) SetTarget(tgt string) *Stats {
	m.target = tgt
	return m
}

// Server ...
func (m *Stats) Server() string {
	return m.server
}

// SetServer ...
func (m *Stats) SetServer(svr string) *Stats {
	m.server = svr
	return m
}

// SetProxy ...
func (m *Stats) SetProxy(svr, tgt string) *Stats {
	m.server = svr
	m.target = tgt
	return m
}

// AddSent ..
func (m *Stats) AddSent(n int64) (new int64) {
	if m.tcp {
		m.rate.AddTCPSent(m.typ, n)
	} else {
		m.rate.AddUDPSent(m.typ, n)
	}
	return atomic.AddInt64(&m.sent, n)
}

// AddRecv ..
func (m *Stats) AddRecv(n int64) (new int64) {
	if m.tcp {
		m.rate.AddTCPRecv(m.typ, n)
	} else {
		m.rate.AddUDPRecv(m.typ, n)
	}
	return atomic.AddInt64(&m.recv, n)
}

// AddSentX ..
func (m *Stats) AddSentX(n int64) (new int64) {
	if m.tcp {
		m.rate.AddTCPSent(m.typ, n)
	} else {
		m.rate.AddUDPSent(m.typ, n)
	}
	return atomic.AddInt64(&m.sentx, n)
}

// AddRecvX ..
func (m *Stats) AddRecvX(n int64) (new int64) {
	if m.tcp {
		m.rate.AddTCPRecv(m.typ, n)
	} else {
		m.rate.AddUDPRecv(m.typ, n)
	}
	return atomic.AddInt64(&m.recvx, n)
}

// Elapsed ..
func (m *Stats) Elapsed() time.Duration {
	return time.Since(m.start)
}

// LogD ..
func (m *Stats) LogD(format string, args ...interface{}) {
	logx.D("%v %v, %v", m.tag, fmt.Sprintf(format, args...), m.Elapsed())
}

// LogI ..
func (m *Stats) LogI(format string, args ...interface{}) {
	logx.I("%v %v, %v", m.tag, fmt.Sprintf(format, args...), m.Elapsed())
}

// LogW ..
func (m *Stats) LogW(format string, args ...interface{}) {
	logx.W("%v %v, %v", m.tag, fmt.Sprintf(format, args...), m.Elapsed())
}

// LogE ..
func (m *Stats) LogE(format string, args ...interface{}) {
	logx.W("%v %v, %v", m.tag, fmt.Sprintf(format, args...), m.Elapsed())
}

// Start ...
func (m *Stats) Start(format string, args ...interface{}) *Stats {
	m.start = time.Now()

	logx.I("%v %v", m.tag, fmt.Sprintf(format, args...))

	if m.tcp {
		m.conn.AddTCP(m.typ, 1)
	} else {
		m.conn.AddUDP(m.typ, 1)
	}
	return m
}

// Done ...
func (m *Stats) Done(format string, args ...interface{}) *Stats {
	if m.tcp {
		m.conn.AddTCP(m.typ, -1)
	} else {
		m.conn.AddUDP(m.typ, -1)
	}

	var logmsg string

	if m.sentx > 0 || m.recvx > 0 {
		logmsg = fmt.Sprintf("%v %v, sent=%v+%v, recv=%v+%v, %v",
			m.tag, fmt.Sprintf(format, args...),
			fx.FormatBytes(m.sentx), fx.FormatBytes(m.sent),
			fx.FormatBytes(m.recvx), fx.FormatBytes(m.recv),
			m.Elapsed())
	} else {
		logmsg = fmt.Sprintf("%v %v, sent=%v, recv=%v, %v",
			m.tag, fmt.Sprintf(format, args...),
			fx.FormatBytes(m.sent), fx.FormatBytes(m.recv),
			m.Elapsed())
	}

	logx.I(logmsg)

	return m
}
